metals/src/main/scala/scala/meta/internal/metals/debug/BuildTargetClassesFinder.scala

```diff
@@ -20,40 +18,46 @@ class BuildTargetClassesFinder(
     index: OnDemandSymbolIndex,
 ) {
 
-  // In case of success returns non-empty list
   def findMainClassAndItsBuildTarget(
       className: String,
-      buildTarget: Option[String],
-  ): Try[List[(b.ScalaMainClass, b.BuildTarget)]] = {
-    findClassAndBuildTarget(
-      className,
-      buildTarget,
-      buildTargetClasses.findMainClassByName(_),
-      buildTargetClasses
+      buildTarget: Option[BuildTarget],
+  ): Either[NoClassError, NonEmptyList[
+    (b.ScalaMainClass, b.BuildTarget)
+  ]] = {
+    val fromBuildTargets = findClassAndBuildTarget[b.ScalaMainClass](
+      className = className,
+      buildTarget = buildTarget,
+      findClassesByName = buildTargetClasses.findMainClassByName(_),
+      classesByBuildTarget = buildTargetClasses
         .classesOf(_)
         .mainClasses
         .values,
-      { clazz: b.ScalaMainClass => clazz.getClassName },
-    ).recoverWith { case ex =>
-      val found = ex match {
+      getClassName = { clazz => clazz.getClassName },
+    )
+
+    fromBuildTargets.left.flatMap { error =>
+      val found = error match {
         // We check whether there is a main in dependencies that is not reported via BSP
         case ClassNotFoundInBuildTargetException(className, target) =>
           revertToDependencies(className, Some(target))
-        case _: ClassNotFoundException =>
+        case _: ClassNotFound =>
           revertToDependencies(className, buildTarget = None)
+        case _ => Nil
       }
       found match {
-        case Nil => Failure(ex)
-        case deps => Success(deps)
+        case Nil => Left(error)
+        case head :: tail => Right(NonEmptyList(head, tail))
```

Comment:
Looks like cats-core have minimal dependencies right now, so this should be ok, Though, if a really small class can be created, I would prefer it, since we would probably not use anything else from cats.